sympy/series/series.py

```diff
@@ -61,3 +63,109 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
     return expr.series(x, x0, n, dir)
+
+
+def inversion_series(f,x,a,n):
+    r"""
+    Computes the Taylor series of the inverse function.
+
+    The formula was proved by Lagrange (1736--1813) and generalized by the German
+    mathematician and teacher Hans Heinrich Bürmann ( --1817), both in the late 18th
+    century.
+
+    Theorem: Suppose x is defined as a function of y by an equation of the form
+
+    f(x) = y
+
+    where f is analytic at a point a and f'(a) is not equal to 0.
+    Then it is possible to invert or solve the equation for x in the form of a series:
+
+    .. math::
+        g(y) = a + \sum_{n = 1}^{oo} \frac{g_n (y - f(a))^n}{n!}
+
+    where
+
+    .. math::
+        g_n = \lim_{y \to a} \left[ \frac{d^{n-1}}{dx^{n-1}}
+        \left( \frac{x-a}{f(x)-f(a)} \right) ^n \right]
+
+    The theorem further states that this series has a non-zero radius of convergence,
+    i.e., x represents an analytic function of y in a neighbourhood of y = f(a).
+
+    Parameters
+    ==========
+
+    f : expression, the expansion of whose inverse is to be found.
+
+    x : variable, in terms of which equation is given, and in terms of which, the
+    expansion will be provided.
+
+    a : point on x-axis in the neighbourhood of which, the inverse expansion is to be found.
+
+    n : no of terms required in the expansion; default value is 3 terms.
+
+    Examples
+    ========
+
+    >>> from sympy.series.series import inversion_series
+    >>> from sympy import exp
+    >>> from sympy.abc import x
+
+    Let the equation whose inverse's expansion is to be found be y = f(x)
+    then the first argument to be supplied must be f(x).
+
+    >>> f = exp(x)
+
+    In this case, the inverse series is that of the logarithmic function around 1 = f(0):
+
+    >>> inversion_series(f,x,0,4)
+    x + (x - 1)**3/3 - (x - 1)**2/2 - 1
+
+    Functions with vanishing derivative cannot be inverted with this method
+
+    >>> f = x**2
+    >>> try:
+    ...    inversion_series(f,x,0,2)
+    ... except ValueError as ex:
+    ...     print(repr(ex))
+    ValueError('The inversion formula requires non-zero derivative.')
+
+    Note
+    ====
+
+    The implementation uses a quasi-newton method to solve an implicit equation.
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Lagrange_inversion_theorem
+    .. [2] http://mathworld.wolfram.com/LagrangeInversionTheorem.html
+    .. [3] http://www.cfm.brown.edu/people/dobrush/am33/Mathematica/lit.html
+       [4] Abramowitz, M. and Stegun, I.A., 1964. "Handbook of Mathematical Functions", Eq. 3.6.6
+    """
+    # reduce to the canonical case a=0,y=0,f'(a)=1
+    f = f.subs({x:x+a})
+    fa =f.subs({x:0})
+    dfa = f.diff(x).subs({x:0})
+    if dfa.equals(0):
+        raise ValueError('The inversion formula requires non-zero derivative.')
+    f = f.series(x,0,n).removeO()
+    # embed into poly ring
+    y = Dummy('y')
+    R = f.as_poly(x).domain.inject(1/dfa)
+    Rxy = R[x,y]
```

Comment:
We need to be careful how we construct this ring. Firstly we should be careful about what it is we are injecting because it can make a nonsensical ring:
```python
In [2]: ZZ.inject(sympify(5))
Out[2]: ZZ[5]
```
Also though I *think* ideally what we want is `K[u,v,x,y]` where:
- `K` is a field (I think this is needed)
- `x` and `y` are the symbols used for the series and inverse series.
- `u` and `v` are symbolic parameters that appear only polynomially in the series coefficients.

It might not be possible to get this form e.g. if the coefficients have negative powers of `u` and `v` in which case we would want something like `K(u, v)[x, y]` with `K(u, v)` being the rational function field `K.frac_field(u, v)`. The code above will create something like `K[u, v][x, y]` rather than `K[u,v,x,y]` or `K(u,v)[x,y]`.

The way to do this is something like:
```python
In [33]: f = x*z + t

In [34]: p = f.as_poly(x)

In [35]: p.domain
Out[35]: ZZ[z,t]

In [36]: p.domain.is_PolynomialRing
Out[36]: True

In [37]: p.domain.domain.get_field().inject(*p.domain.symbols, x, y)
Out[37]: QQ[z,t,x,y]
```
If the domain is not a polynomial ring then we should just require the domain to be a field and then make a polynomial ring over it:
```python
In [40]: (1 + x).as_poly(x).domain.get_field()[x, y]
Out[40]: QQ[x,y]

In [41]: (1 + x/t).as_poly(x).domain.get_field()[x, y]
Out[41]: ZZ(t)[x,y]

In [42]: (cos(t) + x*sin(t)).as_poly(x).domain.get_field()[x, y]
Out[42]: EX[x,y]
```
Ideally there would a ready made function for creating a domain like this.

Reply:
> Also though I _think_ ideally what we want is `K[u,v,x,y]`

We can state this even more precisely:

- the function is expanded and truncated to `f=K[x]` where `R` is some ring. 
- for the inversion formula, we operate in `K[x,y]` where `K=R[1/a]` and `a=f'(0)`.

Reply:
Maybe there should be a general method to extend a domain by the inverse of an element of the domain. It seems like something that would generally be useful rather than going completely to a field e.g. with `ZZ[x]` extending by the inverse of 2 should give `QQ[x]` rather than the field `QQ(x)` (or `ZZ(x)` which is what `ZZ[x].get_field()` returns). Extending by the inverse of a unit can just return the same ring.

Alternatively in some cases you would want to extend only where needed in order to satisfy some division and the question would be if the division can be exact. For example if we are in `ZZ[x]` then it is still possible to compute the division `6/3` or `x/x` etc without needing to go into `QQ[x]` or `QQ(x)`.

Reply:
> extend a domain by the inverse of an element of the domain. It seems like something that would generally be useful rather than going completely to a field

This is exactly what we needed. Yet, what would be the benefit of making this extension accurate - faithful, as opposed to a greedy embedding with EX. Also, recalling what you said once: for performance we may want to disable operations within $K$.

Reply:
We should generally avoid `EX` because it is slow and not always completely well defined. When we can have a well-defined ring that should always be preferred. Likewise if we can get away with a ring rather than a field then that should usually be preferred e.g.:

- `ZZ` is better than `QQ`
- `R[x]` is better than `R(x)`
- `R(x)[y]` is probably better than `R(x, y)`

Reply:
> Alternatively in some cases you would want to extend only where needed in order to satisfy some division and the question would be if the division can be exact. For example if we are in `ZZ[x]` then it is still possible to compute the division `6/3` or `x/x` etc without needing to go into `QQ[x]` or `QQ(x)`.

You are essentially saying that elements of $K[1/a]$ can collapse to just $K$, so sometimes the extension may be not needed. But it is impossible to control, since $1/a$ propagates through calculations making arbitrarily high powers. So with increasing precision you don't have `6/3` but rather `6/27`.
https://mathworld.wolfram.com/SeriesReversion.html